lib/cache/collections.go

```diff
@@ -41,6 +42,88 @@ type collection interface {
 	watchKind() types.WatchKind
 }
 
+// executor[R] is a specific way to run the collector operations that we need
+// for the genericCollector for a generic resource type R.
+type executor[R types.Resource] interface {
+	getAll(ctx context.Context, cache *Cache, loadSecrets bool) ([]R, error)
+	upsert(ctx context.Context, cache *Cache, value R) error
+	deleteAll(ctx context.Context, cache *Cache) error
+	delete(ctx context.Context, cache *Cache, resource types.Resource) error
+}
+
+type genericCollection[R types.Resource, E executor[R]] struct {
+	cache       *Cache
+	watch       types.WatchKind
+	exec        E
+	isSingleton bool
```

Comment:
`executor` must know if it's dealing with a singleton or not, if this is made into an `isSingleton` method in `executor` (which will almost always just return a constant) we don't need to specify this as an additional parameter that we might get wrong when creating a `genericCollection`.